File: apple/Sources/MDEditorUI/Commands.swift
import Foundation

public enum MarkdownCommand: String, Sendable {
    case bold, italic, code, link, heading, bulletList, orderedList, taskList, quote
}

public struct MarkdownCommandResult: Sendable, Equatable {
    public let range: NSRange
    public let text: String
    public let selection: NSRange
}

public func markdownCommand(
    _ command: MarkdownCommand,
    markdown: String,
    selection: NSRange
) -> MarkdownCommandResult {
    let source = markdown as NSString
    let safe = NSIntersectionRange(selection, NSRange(location: 0, length: source.length))
    let selected = source.substring(with: safe)
    func wrap(_ open: String, _ close: String? = nil, placeholder: String = "text")
        -> MarkdownCommandResult {
        let close = close ?? open
        let body = selected.isEmpty ? placeholder : selected
        let openLength = open.utf16.count
        let bodyLength = body.utf16.count
        return MarkdownCommandResult(
            range: safe,
            text: open + body + close,
            selection: NSRange(location: safe.location + openLength, length: bodyLength)
        )
    }
    switch command {
    case .bold: return wrap("**")
    case .italic: return wrap("*")
    case .code: return wrap("`", placeholder: "code")
    case .link: return wrap("[", "](https://)", placeholder: "label")
    default: break
    }
    let lines = source.lineRange(for: safe)
    let body = source.substring(with: lines)
    let parts = body.split(separator: "\n", omittingEmptySubsequences: false)
    let prefixed = parts.enumerated().map { index, line -> String in
        let prefix: String
        switch command {
        case .heading: prefix = "# "
        case .bulletList: prefix = "- "
        case .orderedList: prefix = "\(index + 1). "
        case .taskList: prefix = "- [ ] "
        case .quote: prefix = "> "
        default: prefix = ""
        }
        return prefix + line
    }.joined(separator: "\n")
    let firstPrefix = command == .orderedList ? 3 : [
        MarkdownCommand.heading: 2, .bulletList: 2, .taskList: 6, .quote: 2,
    ][command] ?? 0
    return MarkdownCommandResult(
        range: lines,
        text: prefixed,
        selection: NSRange(
            location: safe.location + firstPrefix,
            length: safe.length + prefixed.utf16.count - lines.length - firstPrefix
        )
    )
}

public extension MarkdownTextView {
    @discardableResult
    func execute(_ command: MarkdownCommand, selection: NSRange? = nil) -> Bool {
        let result = markdownCommand(command, markdown: markdown, selection: selection ?? selectedRange)
        closeUndoGroup()
        guard let textStorage else { return false }
        textStorage.replaceCharacters(in: result.range, with: result.text)
        closeUndoGroup()
        selectedRange = result.selection
        return true
    }
}

File: web/src/commands.ts
import type { MarkdownEditor } from './editor.js';
import type { SelectionRange } from './core.js';

export type MarkdownCommand =
  | 'bold' | 'italic' | 'code' | 'link'
  | 'heading' | 'bullet-list' | 'ordered-list' | 'task-list' | 'quote';

export interface CommandResult {
  start: number;
  end: number;
  text: string;
  selection: SelectionRange;
}

/** Pure source transformation shared by menus, shortcuts, toolbars, and tests. */
export function markdownCommand(
  command: MarkdownCommand,
  markdown: string,
  selection: SelectionRange,
): CommandResult {
  const start = Math.min(selection.start, selection.end);
  const end = Math.max(selection.start, selection.end);
  const selected = markdown.slice(start, end);
  const wrap = (open: string, close = open, placeholder = 'text'): CommandResult => {
    const body = selected || placeholder;
    return {
      start, end, text: open + body + close,
      selection: selected
        ? { start: start + open.length, end: end + open.length }
        : { start: start + open.length, end: start + open.length + body.length },
    };
  };
  if (command === 'bold') return wrap('**');
  if (command === 'italic') return wrap('*');
  if (command === 'code') return wrap('`', '`', 'code');
  if (command === 'link') return wrap('[', '](https://)', 'label');

  const lineStart = markdown.lastIndexOf('\n', Math.max(0, start - 1)) + 1;
  const lineEndCandidate = markdown.indexOf('\n', end);
  const lineEnd = lineEndCandidate < 0 ? markdown.length : lineEndCandidate;
  const lines = markdown.slice(lineStart, lineEnd).split('\n');
  const prefix = command === 'heading' ? '# '
    : command === 'bullet-list' ? '- '
      : command === 'task-list' ? '- [ ] '
        : command === 'quote' ? '> '
          : null;
  const transformed = lines.map((line, index) =>
    command === 'ordered-list' ? `${index + 1}. ${line}` : `${prefix}${line}`,
  ).join('\n');
  const addedToFirst = command === 'ordered-list' ? 3 : prefix!.length;
  return {
    start: lineStart,
    end: lineEnd,
    text: transformed,
    selection: { start: start + addedToFirst, end: end + transformed.length - (lineEnd - lineStart) },
  };
}

export function executeMarkdownCommand(
  editor: MarkdownEditor,
  command: MarkdownCommand,
  selection = editor.selectionRange(),
): boolean {
  if (!selection) return false;
  const result = markdownCommand(command, editor.markdown, selection);
  editor.closeUndoGroup();
  editor.replaceRange(result.start, result.end, result.text);
  editor.closeUndoGroup();
  editor.setSelectionRange(result.selection);
  return true;
}
